### ISSUE TITLE
UCUM service reinitialized for every Context

### ISSUE THREAD

**Issue by csandersdev** [DBCG/cql_engine#505]
While doing some performance evaluation today, I noticed a bunch of stack traces showing up inside [CqlContext.init(Library)](https://github.com/DBCG/cql_engine/blob/50ea3c3720072d77e4a5b445c7586f3e2e1ee4aa/engine/src/main/java/org/opencds/cqf/cql/engine/execution/Context.java#L197) related to the initialization of the UcumEssenceService. This seems like this is a thing that could be initialized once and reuse across contexts to save some cycles.

```
app//org.fhir.ucum.UcumEssenceService.<init>(UcumEssenceService.java:57)
app//org.opencds.cqf.cql.engine.execution.Context.init(Context.java:182)
app//org.opencds.cqf.cql.engine.execution.Context.<init>(Context.java:166)
app//com.ibm.cohort.cql.evaluation.CqlContextFactory.createContext(CqlContextFactory.java:81)
```

**Comment by JPercival** [comment]
I'll have to do some digging to see if the UcumService is thread-safe or not.

**Comment by csandersdev** [comment]
Worst case, we could just store the UcumService in a ThreadLocal static right? I was also wondering if it is something that the user  should configure and provide as input to engine initialization. This would be similar to what is already done in the CqlTranslator.

**Comment by brynrhodes** [comment]
+1 to providing as a parameter, many environments will already have one in context.

**Comment by csandersdev** [comment]
I did some poking around this in the ucum-java code. They don't mention anything about thread safety. I reviewed the code and didn't see anything that would concern me, but it is certainly possible I missed something. I think a better indicator is probably that the HAPI code that uses the UcumEssenseService creates a single static copy for reuse and gets away with it. See https://github.com/hapifhir/hapi-fhir/blob/master/hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/util/UcumServiceUtil.java.
